### ISSUE TITLE
dotnet lambda deploy-function with an updated framework and runtime doesn't propagate the change

### ISSUE THREAD

**Issue by kylegalbraith** [aws/aws-extensions-for-dotnet-cli#149]
## Description
I just went through and upgraded a bunch of my Lambda functions from `dotnetcore2.1` to `dotnetcore3.1`. However, I have a few projects that use the aws-default config file where you specify the runtime and framework. I updated both of those to be on 3.1 but the function did not get updated to the new version.

## Reproduction Steps
1. Have a project that is on `dotnetcore2.1` published to AWS Lambda and setup `aws-lambda-tools-defaults.json` to be like this:
```
{
    "Information" : [
        "This file provides default values for the deployment wizard inside Visual Studio and the AWS Lambda commands added to the .NET Core CLI.",
        "To learn more about the Lambda commands with the .NET Core CLI execute the following command at the command line in the project root directory.",
        "dotnet lambda help",
        "All the command line options for the Lambda command can be specified in this file."
    ],
    "profile"     : "default",
    "region"      : "us-west-2",
    "configuration" : "Release",
    "framework"     : "netcoreapp2.1",
    "function-runtime" : "dotnetcore2.1",
    "function-memory-size" : 256,
    "function-timeout"     : 60,
    "function-handler"     : "UFSX.AddNewRider::UFSX.AddNewRider.Function::FunctionHandler",
    "function-name"        : "UFSX-AddNewRider",
}
```
2. Upgrade the project to `dotnetcore3.1` and update the json file to have `3.1` for `framework` and `function-runtime`.
3. Run `dotnet lambda deploy-function`

Expected: The `deploy-function` should deploy the function code **and** update the function config to be on `3.1`.
Actual: Deployment works but the runtime is not updated.

P.S. I tried the `update-function-config` as well and that didn't change the runtime either.

**Comment by ashishdhingra** [comment]
Hi @kylegalbraith,

Thanks for reporting the issue. Please confirm if you tried using AWS CLI [update-function-configuration](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/lambda/update-function-configuration.html) and if it works.

Thanks,
Ashish

**Comment by kylegalbraith** [comment]
> Please confirm if you tried using AWS CLI [update-function-configuration](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/lambda/update-function-configuration.html) and if it works.

Yes, this works as expected.

**Comment by ashishdhingra** [comment]
Hi @kylegalbraith,

Good afternoon.

Somehow I'm unable to reproduce the issue using `Amazon Lambda Tools for .NET Core applications (4.1.0)`.

**STEPS:**
1. Created Lambda function using .NET CLI using `dotnet new lambda.EmptyFunction`. Here are the contents of various files to target .NET Core 2.1:

- **.csproj file**
```XML
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
    <AWSProjectType>Lambda</AWSProjectType>

    <!-- This property makes the build directory similar to a publish directory and helps the AWS .NET Lambda Mock Test Tool find project dependencies. -->
    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Amazon.Lambda.Core" Version="1.2.0" />
    <PackageReference Include="Amazon.Lambda.Serialization.Json" Version="1.8.0" />
  </ItemGroup>
</Project>
```

- **aws-lambda-tools-defaults.json**
```JSON
{
    "Information" : [
        "This file provides default values for the deployment wizard inside Visual Studio and the AWS Lambda commands added to the .NET Core CLI.",
        "To learn more about the Lambda commands with the .NET Core CLI execute the following command at the command line in the project root directory.",
        "dotnet lambda help",
        "All the command line options for the Lambda command can be specified in this file."
    ],
    "profile"     : "default",
    "region"      : "us-east-2",
    "configuration" : "Release",
    "framework"     : "netcoreapp2.1",
    "function-runtime" : "dotnetcore2.1",
    "function-memory-size" : 256,
    "function-timeout"     : 30,
    "function-handler"     : "AWSLambdaRuntimeUpgradeTest::AWSLambdaRuntimeUpgradeTest.Function::FunctionHandler",
    "function-name"        : "TestLambdaFunction",
    "package-type"         : "Zip",
    "function-role"        : "arn:aws:iam::139480602983:role/lambda_exec_TestLambdaFunction-0",
    "tracing-mode"         : "PassThrough",
    "environment-variables" : "",
    "image-tag"             : ""
}
```

- **Function.cs**
```C#
using Amazon.Lambda.Core;

// Assembly attribute to enable the Lambda function's JSON input to be converted into a .NET class.
[assembly: LambdaSerializer(typeof(Amazon.Lambda.Serialization.Json.JsonSerializer))]

namespace AWSLambdaRuntimeUpgradeTest
{
    public class Function
    {
        
        /// <summary>
        /// A simple function that takes a string and does a ToUpper
        /// </summary>
        /// <param name="input"></param>
        /// <param name="context"></param>
        /// <returns></returns>
        public string FunctionHandler(string input, ILambdaContext context)
        {
            return input?.ToUpper();
        }
    }
}
``` 

2. Deployed the function through .NET CLI using command `dotnet lambda deploy-function`.
```
Amazon Lambda Tools for .NET Core applications (4.1.0)
Project Home: https://github.com/aws/aws-extensions-for-dotnet-cli, https://github.com/aws/aws-lambda-dotnet

Executing publish command
Deleted previous publish folder
... invoking 'dotnet publish', working folder 'C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp2.1\publish'
... Disabling compilation context to reduce package size. If compilation context is needed pass in the "/p:PreserveCompilationContext=false" switch.
... dotnet publish --output "C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp2.1\publish" --configuration "Release" --framework "netcoreapp2.1" /p:GenerateRuntimeConfigurationFiles=true --runtime rhel.7.2-x64 --self-contained false  /p:PreserveCompilationContext=false --manifest "C:\Users\ashdhin\AppData\Local\Temp\tmp885F.tmp"
... publish: Microsoft (R) Build Engine version 16.8.0+126527ff1 for .NET
... publish: Copyright (C) Microsoft Corporation. All rights reserved.
... publish:   Determining projects to restore...
... publish:   Restored C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\AWSLambdaRuntimeUpgradeTest.csproj (in 354 ms).
... publish:   AWSLambdaRuntimeUpgradeTest -> C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp2.1\rhel.7.2-x64\AWSLambdaRuntimeUpgradeTest.dll
... publish:   AWSLambdaRuntimeUpgradeTest -> C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp2.1\publish\
Zipping publish folder C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp2.1\publish to C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp2.1\AWSLambdaRuntimeUpgradeTest.zip
... zipping: Amazon.Lambda.Core.dll
... zipping: Amazon.Lambda.Serialization.Json.dll
... zipping: AWSLambdaRuntimeUpgradeTest.deps.json
... zipping: AWSLambdaRuntimeUpgradeTest.dll
... zipping: AWSLambdaRuntimeUpgradeTest.pdb
... zipping: AWSLambdaRuntimeUpgradeTest.runtimeconfig.json
... zipping: Newtonsoft.Json.dll
Created publish archive (C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp2.1\AWSLambdaRuntimeUpgradeTest.zip).
Creating new Lambda function
New Lambda function created
```
<img width="797" alt="Screen Shot 2020-12-09 at 2 54 09 PM" src="https://user-images.githubusercontent.com/67916761/101700299-a5899d80-3a31-11eb-9bc3-58cdbd618d75.png">

3. Upgraded project to .NET Core 3.1 (refer article [Announcing AWS Lambda support for .NET Core 3.1](https://aws.amazon.com/blogs/compute/announcing-aws-lambda-supports-for-net-core-3-1/)). Here are the contents of various files to target .NET Core 3.1:

- **.csproj file**
```XML
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<TargetFramework>netcoreapp3.1</TargetFramework>
		<GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
		<AWSProjectType>Lambda</AWSProjectType>

		<!-- This property makes the build directory similar to a publish directory and helps the AWS .NET Lambda Mock Test Tool find project dependencies. -->
		<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
	</PropertyGroup>
	<ItemGroup>
		<PackageReference Include="Amazon.Lambda.Core" Version="1.2.0" />
		<PackageReference Include="Amazon.Lambda.Serialization.SystemTextJson" Version="2.1.0" />
	</ItemGroup>
</Project>
```

- **aws-lambda-tools-defaults.json**
```JSON
{
  "Information": [
    "This file provides default values for the deployment wizard inside Visual Studio and the AWS Lambda commands added to the .NET Core CLI.",
    "To learn more about the Lambda commands with the .NET Core CLI execute the following command at the command line in the project root directory.",
    "dotnet lambda help",
    "All the command line options for the Lambda command can be specified in this file."
  ],
  "profile": "default",
  "region": "us-east-2",
  "configuration": "Release",
  "framework": "netcoreapp3.1",
  "function-runtime": "dotnetcore3.1",
  "function-memory-size": 256,
  "function-timeout": 30,
  "function-handler": "AWSLambdaRuntimeUpgradeTest::AWSLambdaRuntimeUpgradeTest.Function::FunctionHandler",
  "function-name": "TestLambdaFunction",
  "package-type": "Zip",
  "function-role": "arn:aws:iam::139480602983:role/lambda_exec_TestLambdaFunction-0",
  "tracing-mode": "PassThrough",
  "environment-variables": "",
  "image-tag": ""
}
```

- **Function.cs**
```C#
using Amazon.Lambda.Core;

// Assembly attribute to enable the Lambda function's JSON input to be converted into a .NET class.
[assembly: LambdaSerializer(typeof(Amazon.Lambda.Serialization.SystemTextJson.DefaultLambdaJsonSerializer))]

namespace AWSLambdaRuntimeUpgradeTest
{
    public class Function
    {

        /// <summary>
        /// A simple function that takes a string and does a ToUpper
        /// </summary>
        /// <param name="input"></param>
        /// <param name="context"></param>
        /// <returns></returns>
        public string FunctionHandler(string input, ILambdaContext context)
        {
            return input?.ToUpper();
        }
    }
}
``` 

4. Re-deployed the function through .NET CLI using command `dotnet lambda deploy-function`.
```
Amazon Lambda Tools for .NET Core applications (4.1.0)
Project Home: https://github.com/aws/aws-extensions-for-dotnet-cli, https://github.com/aws/aws-lambda-dotnet

Executing publish command
Deleted previous publish folder
... invoking 'dotnet publish', working folder 'C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp3.1\publish'
... dotnet publish --output "C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp3.1\publish" --configuration "Release" --framework "netcoreapp3.1" /p:GenerateRuntimeConfigurationFiles=true --runtime linux-x64 --self-contained false
... publish: Microsoft (R) Build Engine version 16.8.0+126527ff1 for .NET
... publish: Copyright (C) Microsoft Corporation. All rights reserved.
... publish:   Determining projects to restore...
... publish:   Restored C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\AWSLambdaRuntimeUpgradeTest.csproj (in 166 ms).
... publish:   AWSLambdaRuntimeUpgradeTest -> C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp3.1\linux-x64\AWSLambdaRuntimeUpgradeTest.dll
... publish:   AWSLambdaRuntimeUpgradeTest -> C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp3.1\publish\
Zipping publish folder C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp3.1\publish to C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp3.1\AWSLambdaRuntimeUpgradeTest.zip
... zipping: Amazon.Lambda.Core.dll
... zipping: Amazon.Lambda.Serialization.SystemTextJson.dll
... zipping: AWSLambdaRuntimeUpgradeTest.deps.json
... zipping: AWSLambdaRuntimeUpgradeTest.dll
... zipping: AWSLambdaRuntimeUpgradeTest.pdb
... zipping: AWSLambdaRuntimeUpgradeTest.runtimeconfig.json
Created publish archive (C:\Users\<user>\source\repos\AWSLambdaRuntimeUpgradeTest\src\AWSLambdaRuntimeUpgradeTest\bin\Release\netcoreapp3.1\AWSLambdaRuntimeUpgradeTest.zip).
Updating code for existing function
Updating runtime configuration for function TestLambdaFunction
```
<img width="797" alt="Screen Shot 2020-12-09 at 2 55 37 PM" src="https://user-images.githubusercontent.com/67916761/101700588-28aaf380-3a32-11eb-88db-a0278927453d.png">


As you could see above, the function runtime is updated to `.NET Core 3.1 (C#/PowerShell)`. Please try using the latest version of Amazon Lambda Tools and see if it resolves the issue.

Thanks,
Ashish

**Comment by AbubakerB** [comment]
I've just ran across this issue now and tracked it down to the update configuration not propagating the handler and runtime changes if `"package-type": "zip"` is not specified in the `aws-lambda-tools-defaults.json` file.

I assume you didn't come across this in your reproduction @ashishdhingra because your config had that property available. If you remove it and change from netcoreapp2.1 to 3.1, the issue occurs.

I have sent over a PR to fix it. For now I've just added `"package-type": "zip"` to my json file until this issue is resolved
